src/chain/validity.rs

```diff
@@ -54,3 +97,367 @@ impl Display for ValidationError {
 }
 
 impl std::error::Error for ValidationError {}
+
+pub fn pre_validate_transaction(
+    txn: &TransactionMessage,
+    block_number: u64,
+    config: &ChainConfig,
+    base_fee_per_gas: Option<U256>,
+) -> Result<(), ValidationError> {
+    let rev = config.revision(block_number);
+
+    if let Some(chain_id) = txn.chain_id() {
+        if rev < Revision::Spurious || chain_id != config.chain_id {
+            return Err(ValidationError::WrongChainId);
+        }
+    }
+
+    match txn.tx_type() {
+        TxType::EIP2930 => {
+            if rev < Revision::Berlin {
+                return Err(ValidationError::UnsupportedTransactionType);
+            }
+        }
+        TxType::EIP1559 => {
+            if rev < Revision::London {
+                return Err(ValidationError::UnsupportedTransactionType);
+            }
+        }
+        TxType::Legacy => {}
+    }
+
+    if let Some(base_fee_per_gas) = base_fee_per_gas {
+        if txn.max_fee_per_gas() < base_fee_per_gas {
+            return Err(ValidationError::MaxFeeLessThanBase);
+        }
+    }
+
+    // https://github.com/ethereum/EIPs/pull/3594
+    if txn.max_priority_fee_per_gas() > txn.max_fee_per_gas() {
+        return Err(ValidationError::MaxPriorityFeeGreaterThanMax);
+    }
+
+    let g0 = intrinsic_gas(txn, rev >= Revision::Homestead, rev >= Revision::Istanbul);
+    if u128::from(txn.gas_limit()) < g0 {
+        return Err(ValidationError::IntrinsicGas);
+    }
+
+    Ok(())
+}
+
+async fn get_parent<'storage, S>(
+    state: &S,
+    header: &BlockHeader,
+) -> anyhow::Result<Option<BlockHeader>>
+where
+    S: State<'storage>,
+{
```

Comment:
Better have something like
```
    if (header.number == 0) {
        return std::nullopt;
    }
```
in the beginning to avoid potential underflow. See https://github.com/torquem-ch/silkworm/pull/415

Reply:
Thanks for the heads up!